### DISCUSSION TITLE
problems with triggering Arduino after condition is met in workflow

### DISCUSSION THREAD

**Discussion by LaurenzMuessig** [bonsai-rx/bonsai#1529] [Q&A] [answered]
Hi guys,

More Arduino related issues from me. Sorry. In the attached workflow I am tracking the position of an animal (using the position of 2 LEDs) while its running around in a square box. Whenever he enters a reward zone (arbitrary circular area) I want to send a signal to my Arduino to initiate the reward delivery. _inRewZone_ generates a Boolean that is true when the animal is inside the reward zone and at the moment I just want to trigger the Arduino whenever the Boolean changes (to check that it works in principle). 
In the workflow I have the Arduino trigger (_SerialWriteLine_) attached to a keypress and that works. However if I attach it to the _DistinctUntilChanged_ branch I get the following error (even though the _String_ node in both cases produces the output aperiodically): 

Bonsai.WorkflowRuntimeException: The given port name does not start with COM/com or does not resolve to a valid serial port.
Parameter name: portName ---> System.ArgumentException: The given port name does not start with COM/com or does not resolve to a valid serial port.
Parameter name: portName
   at System.IO.Ports.SerialStream..ctor(String portName, Int32 baudRate, Parity parity, Int32 dataBits, StopBits stopBits, Int32 readTimeout, Int32 writeTimeout, Handshake handshake, Boolean dtrEnable, Boolean rtsEnable, Boolean discardNull, Byte parityReplace)
   at System.IO.Ports.SerialPort.Open()
   at Bonsai.IO.Ports.SerialPortManager.ReserveConnection(String portName, SerialPortConfiguration serialPortConfiguration)
   at Bonsai.IO.Ports.ObservableSerialPort.<>c__DisplayClass2_0`1.<WriteLine>b__0()
   at System.Reactive.Linq.ObservableImpl.Using`2._.Run()
   --- End of inner exception stack trace ---

Any help/nudge in the right direction is greatly appreciated. Thanks.

Cheers

Laurenz 
[temp.zip](https://github.com/bonsai-rx/bonsai/files/12320944/temp.zip)

**Comment by bruno-f-cruz** [comment, marked as answer]
Hey
Can you try moving the `CreateSerialPort` source to the top of the workflow and give it another try?

**Reply by LaurenzMuessig** [reply]
Hey @bruno-f-cruz. Apologies for the delayed response but I have been away on holiday. That indeed solves the problem. After moving the CreateSerialPort node to the top of the workflow everything seems to work. I wasn't actually aware that the (vertical) order of the nodes in a workflow mattered for stuff like initialising a port. That is good to know. Thanks for your help. 

Cheers

Laurenz
